# Python の基礎 （解答）

まずは簡単な足し算・引き算をしていきましょう。
まずは、足し算、引き算です。数学の記号と同様に足し算には+、引き算には-を使用します。注意点として、プログラミングでは、基本的にすべてのアルファベット、記号には半角を用います。全角を用いるとエラーが起きてしまうのでご注意ください。



jupyter notebookにて以下のようにコードを打ち、「shift + Enter（return）」で実行してください。

## 計算の基礎

### 足し算

In [1]:
10+2

12

### 掛け算・割り算

掛け算、割り算について見ていきましょう。先程までとは違い、数学とは異なる記号を用います。掛け算は*（アスタリスク）、割り算は/（スラッシュ）です。まずは、100 * 1.08 を計算してみましょう。

In [2]:
100 * 1.08 

108.0

### べき乗
最後はべき乗になります。2の3 乗や5の4 乗といったべき乗は**で表現します。


In [3]:
2**3

8

### 変数
プログラミングで非常に重要な概念となる変数についてお伝えします。
何か計算した結果を後々使用するために保持していたり、頻繁に用いるような値を格納しておくために便利なのが変数です。計算結果等を保持するための箱をイメージしていただけるとわかりやすいです。例として、「3」という数値を「a 」という変数に格納する場合を見ていきます。
![変数](./images/hensu.png)


In [4]:
a =3

print(a) で中身を確認します。

In [5]:
print(a)

3


「print( )」は( )内に記述した変数や値を表示するPythonの機能です。Pythonにはこのような機能がたくさんありますが、専門用語で関数と呼びます。非常に重要な概念ですので名前だけでも覚えていただければと思います。
　他の変数も書いてみましょう。10を変数bに代入します


In [6]:
b = 10

　a、bという変数を作成しましたが、実際に変数名には意味のある名前を付けるのが
一般的です。例えば、100 円のノートを購入するとしましょう。（2019 年9月までの）購
入時には消費税8%が加えられるので、実質108 円になります。これをコードで記述すると以下のようなイメージになります。
価格を「price」、消費税を「tax」という変数名にするのは、問題設定と合致していてイメージしやすいですよね。このように、その変数に何が入っているかわかるように変数名を記述するのがポイントです。他の人が自分のコードを見ても、「この変数は〇〇の値が入っているんだね」と理解してもらえるのが理想です。


___
§　ワーク１

①　価格について “price” という名前の変数を作成し 100　を代入します。

②　消費税について”tax”という名前の変数を作成し、 1.08 を代入します。

③　①x②で価格を計算してください。

④　3,900　円の本　と　1,200円の本を消費税10％で購入した場合の金額はいくらですか？　 
___


In [7]:
#価格
price = 3900 + 1200

#消費税
tax =1.1 

In [8]:
price * tax

5610.0

## データ型
コンピューターが扱っているデータというものには型があります。例えるならば、日本語や英語って、大きな括りでいうと言語になりますよね。 この言語の中にも日本語や英語だけではなく、ドイツ語、フランス語という々な「タイプ」（型）があります。それらと同様に、データの中にも数値があったり文字があったりします。数値の中には整数があったり、小数があったりという風に、データも一括りにできるわけではなく、色々な型があります。このようなデータの型のことをデータ型と呼びます。


#### 数値型
　まずは、今まで扱ってきた数値について見ていきましょう。
 
##### 整数
　まずは、数値型の1つである整数についてです。Pythonでは整数のことをint（integer ）でよく表現します。　以下の120という数値のデータ型は整数であり、コード上でデータ型を確認したい際には「type( ) 」を用います。


In [3]:
#a = 120

In [4]:
#type(a)

#### 浮動小数点数（小数）
　数値型の1つである浮動小数点数は、いわゆる小数のことです。浮動小数点数のことをfloatで表現します。整数同様に「type( ) 」でデータ型を確認してみましょう


In [5]:
#a =1.01

In [6]:
#type(a)

#### 文字列型
文字列型はstr（string ） で表現します。


In [7]:
#name = 'bashii'

In [8]:
#type(name)

数値であっても”文字型”に変換することは可能です。

In [9]:
#a= str(a)

In [10]:
#a

In [11]:
#type(a)

## 複数データの扱い
今までは1つの値のみを格納するデータ型についてお伝えしてきましたが、複数の値を扱うデータ型についてもお伝えいたします

### リスト型
1つ目がリスト型です。使い方等の詳細は次節にてお伝えしますが、リスト型を扱う
際は[ ]内に格納したい複数の値を「, 」区切りで入れます。これがリスト型の定義の
仕方になります。


In [13]:
#c = [10,20,30]
#c

In [14]:
#type(c)

### リスト型の扱い
リスト型は、以下のように[ ]内に,区切りで格納したい値を記述します。例として、学校の3 科目のテストの点数、国語50点、数学80点、英語60点の値をリストでまとめるとしましょう。テストの点数をまとめるので変数名をscoresにします。


In [15]:
#scores = [50,80,60]

このリスト（リスト型の変数名のこと）に格納されている各値のことを要素と呼びます
リストの一番目を取り出すには　
scores[0]  と打ちます。
最後は scores[-1] になります。

![scores-list](./images/scores-list.png)

In [16]:
#scores[0:2]

上記は「0 番目の要素から2 番目の要素までを取り出す」という意味に
なりそうですが、実際には「0 番目の要素から2 番目の要素の1つ前の要素までを取り
出す」になります。

したがって、0 番目の要素である50、2 番目の要素の1つ前の要素
つまり1番目の要素である80までを取り出すことを意味します。


まとめると、変数名[ 開始位置: 終了位置の1つ後]で抽出することができます。
また、開始位置の要素番号が0である場合は、以下のように省略することも可能です。


In [17]:
#scores[:2]

以下も同じ要素の抽出方法になります。

In [18]:
#scores[:-1]

___
§　ワーク

①　1月から6月までの名前のリストを作ってください。 -> Jan, Feb, Mar, Apr, May, Jun 

②　1月から3月までを取り出してください

③　6月（最後の一つ以外）を取り出してください
___

In [19]:
#month = ['Jun', 'Feb', 'Mar', 'Apr', 'May', 'Jun']

In [20]:
#type(month)

In [21]:
# 3 から一つ前の 0,1,2 が取得される
#month[0:3]

In [81]:
#0は省略できる
#month[:3]

In [82]:
#最後の一つ　を　除いて・・・という書き方が楽
#month[:-1]

### 辞書型　→　ここはまたあとで必要になったときにやりましょう。
　複数のデータを扱う際に、辞書型というデータ型も存在します。辞書型は、{ }で定義します。こちらは魏リッドサーチという最適な機械学習のハイパーパラメータを探す際に利用しますので、またその時に思い出しましょう。ここではデータ型の1つとして存在することを覚えておいてください。

In [22]:
#辞書型の練習
#person = {'firstname':'iwahashi','lastname':'tomohiro','nickname':'bashii'}

In [23]:
#person["nickname"]

In [24]:
# 辞書型については .items を使って以下のような使い方ができます。
#for key,value in person.items():
#    print('私の' + '{}'.format(key) + 'は、'+ '{}'.format(value) +'です。')


## Numpyの利用
数学的な計算を行う場合NumpyのArray形にします。

List型どうしを足し算してみましょう。例えば

hundred = [100,200,300]

ten = [10,20,30]

hundred + ten


In [25]:
#hundred = [100,200,300]

In [26]:
#ten = [10,20,30]

In [27]:
#hundred + ten

In [28]:
#type(hundred)

思った通りに行きませんね・・・　数学的な計算（行列計算）はNumpyを利用します。

In [37]:
# 必要なパッケージは import します。→　とても重要です！
# as xx と名前を付けます。

#import numpy as np

In [30]:
# np.array で Numpy 型に変換します。
#hundred_a = np.array(hundred)

In [31]:
#type(hundred_a)
# ndarray に変わりましたね。

In [33]:
# こちらもnp.array で Numpy 型に変換します。
#ten_a =np.array(ten)

In [34]:
#足し算してみましょう
#hundred_a +  ten_a

### 補足

※　ここは補足です：興味のある方だけ確認ください。　

―　Numpyを使った行列計算について　ー　

参考：　https://teratail.com/questions/234871

In [38]:
# Nunpy をつかうと行列計算ができるのです。
A = np.asarray([[1,2],[3,4]])

In [42]:
# .Tは転置という意味です。
#x = np.asarray([[1,2]]).T

In [43]:
# dot はベクトルの掛け算です。
#A.dot(x)

In [46]:
#逆行列も計算できます。
#A_inv = np.linalg.inv(A)

In [47]:
#A_inv

In [48]:
#逆行列はかけたら単位行列になるのでしたね！？　→　計算がデジタルのため若干０になりません。
#A.dot(A_inv)

Numpyを使わなくても、FOR文でいけるかもしれませんね。

In [49]:
#h_t = zip(hundred,ten)
#for i in h_t:
#    print (i[0] + i[1])

## Pandas の利用方法

### pandasとは
　pandas（パンダス）とは、Pythonの外部パッケージです。パッケージとは、簡単にお伝えすると「便利な機能をあつめたツール群」のことを指します。他の言語ではライブラリと呼んだりもします。Pythonには、pandasの他にも様々なパッケージがあり、様々な用途に応じたパッケージを使用することで目的を素早く終える手助けをしてくれます。
 


In [55]:
#モジュールを読み込む際にはimport文を使用します。import モジュール名で読み込むことができます
import pandas as pd

### DataFrame型
　次にDataFrame型を見ていきましょう。「pd.DataFrame(2次元データ) 」で定義することができます。「df」（dataframeの略）という変数で定義しましょう。

In [56]:
df = pd.DataFrame([[1,2,3],[4,5,6],[7,8,9]])

In [57]:
df

0  1  2
0  1  2  3
1  4  5  6
2  7  8  9

上記の( ) 内が少しわかりにくいですが、以下のようにリストを二重で用いて2 次元
データを表現しています。

![2次元データの作成](images/pandas-1.png)

type()でDataFrame型になっていることを確認しましょう。

In [58]:
type(df)

pandas.core.frame.DataFrame

### csvファイルの読み込み、書き出し
DataFrame型ともに定義の仕方をお伝えしましたが、先程のように自ら値を入れて定義するというのは稀です。 実際は、csvファイルやデータベース等からpandasの形式でデータを読み込むことが一般的です。今回はcsvファイルからのデータ読み込み、書き出しについてお伝えします。

In [59]:
ls

 ドライブ C のボリューム ラベルは Windows です
 ボリューム シリアル番号は 2CB0-A238 です

 C:\Users\USER\Documents\iwahashi\iwahashi\splineglobal\TableauDS学習塾\long3\week3\tableau4ds_work のディレクトリ

2020/05/26  15:05    <DIR>          .
2020/05/26  15:05    <DIR>          ..
2020/05/26  14:56    <DIR>          .ipynb_checkpoints
2020/03/18  20:27           100,820 【ワーク】Tableauで始めるデータサイエンス学習塾-WEEK3-Pythonの基礎_20200318.ipynb
2020/03/18  21:14           162,588 【ワーク】Tableauで始めるデータサイエンス学習塾-WEEK3-機械学習の基礎_20200318.ipynb
2020/05/26  15:05            75,851 【解答】Tableauで始めるデータサイエンス学習塾-WEEK3-Pythonの基礎_20200318.ipynb
2020/03/23  21:19           149,892 【解答】Tableauで始めるデータサイエンス学習塾-WEEK3-機械学習の基礎_20200318.ipynb
2020/03/18  01:13            87,451 Chapter10.ipynb
2020/05/26  14:56    <DIR>          data
2020/03/18  03:24               908 export.csv
2020/03/18  20:16               908 export_iwahashi.csv
2020/03/17  16:29           133,059 housing_tokyo_chuo_line_week3.csv
2020/05/26  14:56    <DIR>          images
2020/03/22  18:17  

pd.read_csv（ファイルまでのパス）でcsvを読み込むことができますが、今回は実行ファイルとcsvファイルが同
フォルダ内に格納されているため、ファイル名の指定のみで読み込むことができます。 "housing_tokyo_chuo_line_week3.csv" を読み込みます。

In [60]:
df =  pd.read_csv("housing_tokyo_chuo_line_week3.csv")

In [61]:
#中身を確認します
df

uid  years  minutes  sqrm   distance  renovate  express     price
0       13101-1     34        3    35   2.983801         0        0  24000000
1       13101-2     37        4    20   2.983801         0        0  13000000
2       13101-4     34        3    30   2.983801         0        0  22000000
3       13101-5     35        1    70   2.983801         0        0  69000000
4       13101-6     34        3    25   2.983801         0        0  21000000
...         ...    ...      ...   ...        ...       ...      ...       ...
2994  13214-847     19        4    20  25.956039         0        1  17000000
2995  13214-848     21        1    80  25.956039         0        1  65000000
2996  13214-849     21        1    65  25.956039         0        1  42000000
2997  13214-850     15        3    25  25.956039         0        1  19000000
2998  13214-851     14        2    20  25.956039         0        1  14000000

[2999 rows x 8 columns]

In [62]:
#頭の数行をとるのは head() を使います
df.head()

uid  years  minutes  sqrm  distance  renovate  express     price
0  13101-1     34        3    35  2.983801         0        0  24000000
1  13101-2     37        4    20  2.983801         0        0  13000000
2  13101-4     34        3    30  2.983801         0        0  22000000
3  13101-5     35        1    70  2.983801         0        0  69000000
4  13101-6     34        3    25  2.983801         0        0  21000000

In [63]:
#頭の20行をとってexport_dataに書き戻します。
export_data = df.head(20)

### csvファイルへ書き出し
　今作成したexport_dataをcsvに書き出してみましょう。書き出しの際は「to_csv(書き出しファイル名) 」を用います。一番左に表示されている行番号を示すインデック
スも書き出されないように「index=False」を指定する必要があります。

In [64]:
export_data.to_csv("export.csv", index=False)

書き出しに成功すれば、作業フォルダ内にexport.csvというファイルが作成されて
いるかと思います。

In [65]:
#ls

**memo:**

• リスト型：データの追加やデータの削除等のシンプルなデータ操作に用い
る。多様な機能を持っているわけではないため、複雑な処理には不向き。

• NumPy：線形代数演算等の数学的な計算を必要とする際に用いる。内部
はC言語で実装されており、計算速度が速い。

• pandas：データを結合したり、データの条件抽出をしたり、複雑なデータ
整理を行う場合に用いる。NumPyをベースに作られており、多様なデー
タ形式に対応できる。

### 条件抽出
部分抽出
　pandasで読み込んだデータの一部を抽出する方法についてお伝えいたします。方法は複数ありますが、カラム名を指定した方法と、iloc[ ]を用いた方法の2つについてお伝えします。例として、先頭2 列や5 列目のみのデータを抽出する方法について見ていきましょう。

#### まずはカラム名を指定する方法です。

In [66]:
df[['years','minutes']]

years  minutes
0        34        3
1        37        4
2        34        3
3        35        1
4        34        3
...     ...      ...
2994     19        4
2995     21        1
2996     21        1
2997     15        3
2998     14        2

[2999 rows x 2 columns]

しっかりと先頭2 列を抽出することができました。注意点として、2次元データを取
得するため、[ ]が二重になっています。[ ]が1つだと1次元データを取り出すことになっ
てしまい、エラーが起きてしまいます。

#### iloc[ ]を用いた方法
　カラム名を指定した方法の場合、複数列抽出する際にすべてのカラム名を記載する必要があります。10 列取得したい場合、10 列分のカラム名を[ ] 内に記載する必要があり、非常に大変です。そこで登場するのがiloc[ ]です。iloc[ ]を用いた抽出方法のイメージは以下のようになります。
 
![iloc](images/iloc.png)

早速、先程と同じ列を抽出してみましょう。今回抽出したい行は全てなので、[ ] 内
に:と記述します。「: 」は最初から最後まで全てを意味します。列は0 列目から1列目ま
でを取得したいので「0:2」となります。指定の仕方はリストと同様です。

In [67]:
df.iloc[:,1:3]

years  minutes
0        34        3
1        37        4
2        34        3
3        35        1
4        34        3
...     ...      ...
2994     19        4
2995     21        1
2996     21        1
2997     15        3
2998     14        2

[2999 rows x 2 columns]

---
### §ワーク　"uid" 　と　"price" 以外を取り出して、x という変数に入れてみてましょう。
---


In [68]:
x = df.iloc[:,1:-1]

In [69]:
x

years  minutes  sqrm   distance  renovate  express
0        34        3    35   2.983801         0        0
1        37        4    20   2.983801         0        0
2        34        3    30   2.983801         0        0
3        35        1    70   2.983801         0        0
4        34        3    25   2.983801         0        0
...     ...      ...   ...        ...       ...      ...
2994     19        4    20  25.956039         0        1
2995     21        1    80  25.956039         0        1
2996     21        1    65  25.956039         0        1
2997     15        3    25  25.956039         0        1
2998     14        2    20  25.956039         0        1

[2999 rows x 6 columns]

---
### §ワーク "price" だけを取り出して、ｔ という変数に入れてみてましょう。
---


In [70]:
t = df.iloc[:,-1]

iloc[ ]を用いてxとtに分けました。早速、両者を結合（concatenate）し、元の
DataFrameに戻しましょう。DataFrameの結合には「pd.concat( ) 」を使用します。
今回は横方向に結合するため、「axis=1」という値も指定する必要があります。

In [71]:
pd.concat([x,t],axis=1)

years  minutes  sqrm   distance  renovate  express     price
0        34        3    35   2.983801         0        0  24000000
1        37        4    20   2.983801         0        0  13000000
2        34        3    30   2.983801         0        0  22000000
3        35        1    70   2.983801         0        0  69000000
4        34        3    25   2.983801         0        0  21000000
...     ...      ...   ...        ...       ...      ...       ...
2994     19        4    20  25.956039         0        1  17000000
2995     21        1    80  25.956039         0        1  65000000
2996     21        1    65  25.956039         0        1  42000000
2997     15        3    25  25.956039         0        1  19000000
2998     14        2    20  25.956039         0        1  14000000

[2999 rows x 7 columns]

# 以上で終了です。お疲れさまでした！

# おまけ

### for文は使う機会が多いです。データの前処理、繰り返し処理など、細かい処理はPythonで書いてしまった方が早い場合があります。さっとかけると、とても役に立ちます。

In [119]:
numbers = [1,2,3,4,5,6,7]

In [123]:
# For 分の基礎
for num in numbers:
    print(num**2)
    

1
4
9
16
25
36
49


### ワーク: 以下　水星、金星、地球、火星、木星の半径(km)があるときの、それぞれの惑星の体積はいくつですか？

In [129]:
# 体積の公式は＝　4/3 * pi * r**2  (四分の三 π R二乗)
# πは math.pi 
import math
planet_r = [2439.7, 6051.8, 6371.0, 3390.0, 69911 ]

In [144]:
# For文を使ったとき

In [145]:
planet_volume = []
for r in  planet_r:
    volume = 4/3*math.pi*r**2
    print(volume)
    planet_volume.append(volume)

24932249.351347562
153411438.89303437
170021490.63659608
48137995.91242568
20472912856.908886


In [146]:
# arrayを使ったとき

In [143]:
planet_r_a = np.array(planet_r)
planet_volume_a = 4/3*math.pi*planet_r_a**2
planet_volume_a

array([2.49322494e+07, 1.53411439e+08, 1.70021491e+08, 4.81379959e+07,
       2.04729129e+10])

### ヘッダーに余計な列が入っているファイルを何とかする方法
### ヘッダーが2行になっていたり余計な行が入っていると、BI製品はあちゃーとなります。そんな時、pythonで整形してしまうという手があります。

In [72]:
df = pd.read_csv('./data/売上/売上東京202004.csv')

In [73]:
df

#2020年４月の売上 Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4
0           日付         地区       顧客CD       商品CD         個数
1     2020/4/1         東京        C01        P01         10
2     2020/4/2         東京       C-01        P02         20
3     2020/4/3         東京       C001        P03         30
4     2020/4/4         東京        C02        P01         40
5     2020/4/5         東京        C03        P02         50
6     2020/4/6         東京        C02        P03         60
7     2020/4/7         東京        C01        P01         70
8     2020/4/8         東京        C03        P02         80
9     2020/4/9         東京        C02        P03         90
10   2020/4/10         東京        C01        P01        100

In [75]:
# 列名だけ取り出す。
columname = list(df.iloc[0,])
columname

['日付', '地区', '顧客CD', '商品CD', '個数']

In [225]:
# 本当に必要なデータ部分を抜き出す
df = df.iloc[1:,:]
df

#2020年４月の売上 Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4
1     2020/4/1         東京        C01        P01         10
2     2020/4/2         東京       C-01        P02         20
3     2020/4/3         東京       C001        P03         30
4     2020/4/4         東京        C02        P01         40
5     2020/4/5         東京        C03        P02         50
6     2020/4/6         東京        C02        P03         60
7     2020/4/7         東京        C01        P01         70
8     2020/4/8         東京        C03        P02         80
9     2020/4/9         東京        C02        P03         90
10   2020/4/10         東京        C01        P01        100

In [226]:
# 列名を変更する
df.columns=columname

In [230]:
#ファイルについては出来上がり。
df

日付  地区  顧客CD 商品CD   個数
1    2020/4/1  東京   C01  P01   10
2    2020/4/2  東京  C-01  P02   20
3    2020/4/3  東京  C001  P03   30
4    2020/4/4  東京   C02  P01   40
5    2020/4/5  東京   C03  P02   50
6    2020/4/6  東京   C02  P03   60
7    2020/4/7  東京   C01  P01   70
8    2020/4/8  東京   C03  P02   80
9    2020/4/9  東京   C02  P03   90
10  2020/4/10  東京   C01  P01  100

### ディレクトリのファイルに対してまとめて処理する場合


In [76]:
# ファイル名の取得
import glob
import os

files = []

for f in glob.glob('./data/売上/*.csv'):
    print(os.path.split(f)[1])
    files.append(f)

売上東京202004.csv
売上東京202005.csv
売上東京202006.csv


In [77]:
files

['./data/売上\\売上東京202004.csv',
 './data/売上\\売上東京202005.csv',
 './data/売上\\売上東京202006.csv']

In [78]:
# まずは一つのファイルでやってみる
filepath = files[0]
pd.read_csv(filepath)
df = pd.read_csv(filepath)

filename = os.path.split(filepath)[1]
print(filename)

# 列名だけ取り出す。
columname = list(df.iloc[0,])

#本当に必要なデータ部分を抜き出す
df = df.iloc[1:,:]

# 列名を変更する
df.columns=columname

# 書き出す
df.to_csv('./data/売上整形/'+filename , index=False)


売上東京202004.csv


In [79]:
#for文で回してみる (からっぽで)

for filepath in files:
    print(filepath)

./data/売上\売上東京202004.csv
./data/売上\売上東京202005.csv
./data/売上\売上東京202006.csv


In [80]:
#for文で回してみる　（中身を入れてみる）

for filepath in files:
    #print(filepath)
    
    #filepath = files[0]
    
    pd.read_csv(filepath)
    df = pd.read_csv(filepath)

    filename = os.path.split(filepath)[1]
    #print(filename)

    # 列名だけ取り出す。
    columname = list(df.iloc[0,])

    #本当に必要なデータ部分を抜き出す
    df = df.iloc[1:,:]

    # 列名を変更する
    df.columns=columname

    # 書き出す
    df.to_csv('./data/売上整形/'+filename , index=False)